munch/__init__.py

```diff
@@ -397,13 +397,41 @@ def munchify(x, factory=Munch):
 
         nb. As dicts are not hashable, they cannot be nested in sets/frozensets.
     """
+    return cycle_munchify(x, factory, dict())
+
+# Munchify x, use seen to track object cycles
+def cycle_munchify(x, factory, seen):
+    try:
+        return seen[id(x)]
+    except KeyError:
+        pass
+
+    m = seen[id(x)] = cycle_premunchify(x, factory, seen)
+    cycle_postmunchify(m, x, factory, seen)
+    return m
+
+def cycle_premunchify(x, factory, seen):
+    # Here we return a skeleton of munchified x, which is enough to save for later (in case we need to break cycles)
+    # but it needs to revisited later (to complete filling out any cycles we may have encountered)
     if isinstance(x, Mapping):
-        return factory((k, munchify(x[k], factory)) for k in iterkeys(x))
-    elif isinstance(x, (list, tuple)):
-        return type(x)(munchify(v, factory) for v in x)
+        return factory({})
+    elif isinstance(x, list):
+        return type(x)()
+    elif isinstance(x, tuple):
+        return type(x)(cycle_munchify(v, factory, seen) for v in x)
     else:
         return x
 
+def cycle_postmunchify(m, x, factory, seen):
+    # Here we finish munchifying the parts of x that were skeletonized by cycle_premunchify (because they might participate in a cycle)
+    if isinstance(x, Mapping):
+        m.update((k, cycle_munchify(x[k], factory, seen)) for k in iterkeys(x))
+    elif isinstance(x, list):
+        m.extend(cycle_munchify(v, factory, seen) for v in x)
+    elif isinstance(x, tuple):
+        for (mv, v) in zip(m, x):
```

Comment:
please give better names to these... Too many variables spread around with single letters is very confusing to follow

Reply:
This comment is still relevant

Reply:
Should be better / good now